Update gevent

LefterisJP opened the pull request.

LefterisJP commented:
Hmmm ... okay this may need more work.

```
Current thread 0x00007fa0712f6740 (most recent call first):
  File "/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/site-packages/web3/_utils/request.py", line 37 in make_post_request
  File "/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/site-packages/web3/providers/rpc.py", line 95 in make_request
  File "/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/site-packages/web3/providers/base.py", line 103 in isConnected
  File "/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/site-packages/web3/main.py", line 257 in isConnected
  File "/home/runner/work/rotki/rotki/rotkehlchen/chain/ethereum/manager.py", line 304 in attempt_connect
/home/runner/work/_temp/a1fa657c-00c7-4b8a-bbf9-d0fb40c3698c.sh: line 6: 25762 Segmentation fault      (core dumped) python pytestgeventwrapper.py $COVERAGE_ARGS rotkehlchen/tests
```

LefterisJP commented:
With 20.6.2 we get a segfault at the very beginning of all api tests
![2021-01-20_00-01](https://user-images.githubusercontent.com/1658405/105104522-b3abee80-5ab2-11eb-96f1-8f08460bf0bd.png)

LefterisJP commented:
Also tested with the old web3 and it still happens so it's not something to do purely with the web3 library upgrade

LefterisJP commented:
I managed to reproduce locally. As usual by running all api tests in the api directory. But at least at the time of the run I had only the following tests in the directory:

```
-rw-r--r--  1 lefteris users 21483 Jan 20 16:46 test_settings.py
-rw-r--r--  1 lefteris users 14946 Jan 20 15:05 test_statistics.py
-rw-r--r--  1 lefteris users  3512 Jan 20 12:52 test_substrate_manager.py
-rw-r--r--  1 lefteris users 20438 Jan 20 12:15 test_tags.py
-rw-r--r--  1 lefteris users 26082 Jan 20 12:32 test_trades.py
-rw-r--r--  1 lefteris users 51055 Jan 20 12:19 test_uniswap.py
-rw-r--r--  1 lefteris users 31611 Jan 20 12:01 test_users.py
-rw-r--r--  1 lefteris users 22175 Jan 20 11:56 test_yearn_vaults.py
```

I removed all and started adding them backwards in alphabetic order.

It happened at test_users.py and presumably at the same point as shown in the CI:

```
rotkehlchen/tests/api/test_users.py Rotki API server is running at: 127.0.0.1:29920
[20/01/2021 16:52:33 CET] INFO rotkehlchen.api.server: Rotki API server is running at: 127.0.0.1:29920
[20/01/2021 16:52:33 CET] INFO rotkehlchen.rotkehlchen: Unlocking user user=testuser, create_new=True, sync_approval=no, initial_settings=None
[20/01/2021 16:52:34 CET] DEBUG rotkehlchen.accounting.accountant: Initializing Accountant
[20/01/2021 16:52:34 CET] DEBUG rotkehlchen.exchanges.manager: Initializing exchanges
[20/01/2021 16:52:34 CET] DEBUG rotkehlchen.chain.ethereum.manager: Initializing Ethereum Manager with own rpc endpoint: None
[20/01/2021 16:52:34 CET] DEBUG rotkehlchen.chain.substrate.manager: Initializing Kusama manager
[20/01/2021 16:52:34 CET] WARNING rotkehlchen.chain.substrate.manager: Kusama manager won't attempt to connect to nodes connect_at_start=(), connect_on_startup=False, own_rpc_endpoint=None
[20/01/2021 16:52:34 CET] DEBUG rotkehlchen.chain.manager: Initializing ChainManager
[20/01/2021 16:52:34 CET] DEBUG rotkehlchen.rotkehlchen: User unlocking complete
[20/01/2021 16:52:34 CET] DEBUG rotkehlchen.api.rest: Request successful response={"result": {"testuser": "loggedin", "another_user": "loggedout"}, "message": ""}, status_code=200
[20/01/2021 16:52:34 CET] INFO rotkehlchen.api.server.pywsgi: 127.0.0.1 - - "GET /api/1/users HTTP/1.1" 200 216 0.001270
.[20/01/2021 16:52:34 CET] INFO rotkehlchen.rotkehlchen: Logging out user user=testuser
[20/01/2021 16:52:34 CET] INFO rotkehlchen.rotkehlchen: User successfully logged out user=testuser
[20/01/2021 16:52:34 CET] DEBUG rotkehlchen.api.rest: Waiting for greenlets
[20/01/2021 16:52:34 CET] DEBUG rotkehlchen.api.rest: Waited for greenlets. Killing all other greenlets
[20/01/2021 16:52:34 CET] DEBUG rotkehlchen.api.rest: Greenlets killed. Killing zerorpc greenlet
[20/01/2021 16:52:34 CET] DEBUG rotkehlchen.api.rest: Shutdown completed
Rotki API server is running at: 127.0.0.1:29921
[20/01/2021 16:52:34 CET] INFO rotkehlchen.api.server: Rotki API server is running at: 127.0.0.1:29921
[20/01/2021 16:52:35 CET] DEBUG rotkehlchen.api.rest: Request successful response={"result": {"testuser": "loggedout", "another_user": "loggedout"}, "message": ""}, status_code=200
[20/01/2021 16:52:35 CET] INFO rotkehlchen.api.server.pywsgi: 127.0.0.1 - - "GET /api/1/users HTTP/1.1" 200 217 0.000914
.[20/01/2021 16:52:35 CET] DEBUG rotkehlchen.api.rest: Waiting for greenlets
[20/01/2021 16:52:35 CET] DEBUG rotkehlchen.api.rest: Waited for greenlets. Killing all other greenlets
[20/01/2021 16:52:35 CET] DEBUG rotkehlchen.api.rest: Greenlets killed. Killing zerorpc greenlet
[20/01/2021 16:52:35 CET] DEBUG rotkehlchen.api.rest: Shutdown completed
Rotki API server is running at: 127.0.0.1:29922
[20/01/2021 16:52:35 CET] INFO rotkehlchen.api.server: Rotki API server is running at: 127.0.0.1:29922
[20/01/2021 16:52:35 CET] INFO rotkehlchen.rotkehlchen: Unlocking user user=hania, create_new=True, sync_approval=yes, initial_settings=None
[20/01/2021 16:52:35 CET] DEBUG rotkehlchen.accounting.accountant: Initializing Accountant
[20/01/2021 16:52:35 CET] DEBUG rotkehlchen.exchanges.manager: Initializing exchanges
[20/01/2021 16:52:35 CET] DEBUG rotkehlchen.chain.ethereum.manager: Initializing Ethereum Manager with own rpc endpoint: http://localhost:8545
[20/01/2021 16:52:35 CET] DEBUG rotkehlchen.chain.substrate.manager: Initializing Kusama manager
[20/01/2021 16:52:35 CET] WARNING rotkehlchen.chain.substrate.manager: Kusama manager won't attempt to connect to nodes connect_at_start=(<KusamaNodeName.OWN: 0>, <KusamaNodeName.PARITY: 1>), connect_on_startup=False, own_rpc_endpoint=http://localhost:9933
[20/01/2021 16:52:35 CET] DEBUG rotkehlchen.chain.manager: Initializing ChainManager
[20/01/2021 16:52:35 CET] DEBUG rotkehlchen.rotkehlchen: User unlocking complete
[20/01/2021 16:52:35 CET] DEBUG rotkehlchen.api.rest: Request successful response={"result": {"exchanges": [], "settings": {"have_premium": false, "version": 23, "last_write_ts": 0, "premium_should_sync": false, "include_crypto2crypto": true, "anonymized_logs": false, "last_data_upload_ts": 0, "ui_floating_precision": 2, "taxfree_after_period": 31536000, "balance_save_frequency": 24, "include_gas_costs": true, "eth_rpc_endpoint": "http://localhost:8545", "ksm_rpc_endpoint": "http://localhost:9933", "main_currency": "USD", "date_display_format": "%d/%m/%Y %H:%M:%S %Z", "last_balance_save": 0, "submit_usage_analytics": true, "kraken_account_type": "starter", "active_modules": ["makerdao_dsr", "makerdao_vaults", "aave", "compound", "yearn_vaults", "uniswap", "adex"], "frontend_settings": "", "account_for_assets_movements": true, "btc_derivation_gap_limit": 20, "calculate_past_cost_basis": true}}, "message": ""}, status_code=200
[20/01/2021 16:52:35 CET] INFO rotkehlchen.api.server.pywsgi: 127.0.0.1 - - "PUT /api/1/users HTTP/1.1" 200 976 0.298643
[20/01/2021 16:52:35 CET] DEBUG web3.providers.HTTPProvider: Making request HTTP. URI: http://localhost:8545, Method: web3_clientVersion
[20/01/2021 16:52:35 CET] DEBUG web3.providers.HTTPProvider: Making request HTTP. URI: https://api.mycryptoapi.com/eth, Method: web3_clientVersion
[20/01/2021 16:52:35 CET] DEBUG web3.providers.HTTPProvider: Making request HTTP. URI: https://mainnet-nethermind.blockscout.com/, Method: web3_clientVersion
[20/01/2021 16:52:35 CET] DEBUG web3.providers.HTTPProvider: Making request HTTP. URI: https://web3.1inch.exchange, Method: web3_clientVersion
[20/01/2021 16:52:35 CET] DEBUG web3.providers.HTTPProvider: Making request HTTP. URI: https://mainnet.eth.cloud.ava.do/, Method: web3_clientVersion
[20/01/2021 16:52:35 CET] DEBUG web3.providers.HTTPProvider: Making request HTTP. URI: https://web3.1inch.exchange, Method: web3_clientVersion
[20/01/2021 16:52:35 CET] DEBUG web3.providers.HTTPProvider: Making request HTTP. URI: https://nodes.mewapi.io/rpc/eth, Method: web3_clientVersion
[20/01/2021 16:52:35 CET] DEBUG web3.providers.HTTPProvider: Making request HTTP. URI: https://cloudflare-eth.com/, Method: web3_clientVersion
[20/01/2021 16:52:35 CET] WARNING rotkehlchen.user_messages: You do not have an Etherscan API key configured. Rotki etherscan queries will still work but will be very slow. If you are not using your own ethereum node, it is recommended to go to https://etherscan.io/register, create an API key and then input it in the external service credentials setting of Rotki
[20/01/2021 16:52:35 CET] DEBUG rotkehlchen.externalapis.etherscan: Querying etherscan: https://api.etherscan.io/api?module=proxy&action=eth_call&to=0x3d9819210A31b4961b30EF54bE2aeD79B9c9Cd3B&data=0xbb82aa5e
Fatal Python error: Segmentation fault

Thread 0x00007f9368b6d640 (most recent call first):
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/gevent/_threading.py", line 68 in wait
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/gevent/_threading.py", line 158 in get
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/gevent/threadpool.py", line 187 in run

Thread 0x00007f936a408640 (most recent call first):
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/gevent/_threading.py", line 68 in wait
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/gevent/_threading.py", line 158 in get
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/gevent/threadpool.py", line 187 in run

Current thread 0x00007f9376a5c740 (most recent call first):
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/web3/_utils/request.py", line 31 in _get_session
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/web3/_utils/request.py", line 37 in make_post_request
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/web3/providers/rpc.py", line 95 in make_request
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/web3/providers/base.py", line 103 in isConnected
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/web3/main.py", line 257 in isConnected
  File "/home/lefteris/w/rotkehlchen/rotkehlchen/chain/ethereum/manager.py", line 304 in attempt_connect
[1]    33392 segmentation fault (core dumped)  python pytestgeventwrapper.py -xs rotkehlchen/tests/api --pdb
```

LefterisJP commented:
It happened when I added test_settings.py to the mix so I removed the extra and it happened again only with

```
-rw-r--r--  1 lefteris users 21483 Jan 20 17:12 test_settings.py
-rw-r--r--  1 lefteris users 51055 Jan 20 12:19 test_uniswap.py
-rw-r--r--  1 lefteris users 31611 Jan 20 12:01 test_users.py
```

LefterisJP commented:
Only once did I manage to actually not get a segfault but an actual exception that looks similar to the segfault.

```
[20/01/2021 18:48:17 CET] DEBUG web3.providers.HTTPProvider: Making request HTTP. URI: http://localhost:8545, Method: web3_clientVersion 
[20/01/2021 18:48:17 CET] DEBUG web3.providers.HTTPProvider: Making request HTTP. URI: https://api.mycryptoapi.com/eth, Method: web3_clientVersion
[20/01/2021 18:48:17 CET] DEBUG web3.providers.HTTPProvider: Making request HTTP. URI: https://mainnet-nethermind.blockscout.com/, Method: web3_clientVersion
[20/01/2021 18:48:17 CET] DEBUG web3.providers.HTTPProvider: Making request HTTP. URI: https://web3.1inch.exchange, Method: web3_clientVersion
[20/01/2021 18:48:17 CET] DEBUG web3.providers.HTTPProvider: Making request HTTP. URI: https://mainnet.eth.cloud.ava.do/, Method: web3_clientVersion
[20/01/2021 18:48:17 CET] DEBUG web3.providers.HTTPProvider: Making request HTTP. URI: https://web3.1inch.exchange, Method: web3_clientVersion
[20/01/2021 18:48:17 CET] DEBUG web3.providers.HTTPProvider: Making request HTTP. URI: https://nodes.mewapi.io/rpc/eth, Method: web3_clientVersion
[20/01/2021 18:48:17 CET] DEBUG web3.providers.HTTPProvider: Making request HTTP. URI: https://cloudflare-eth.com/, Method: web3_clientVersion
[20/01/2021 18:48:17 CET] WARNING rotkehlchen.user_messages: You do not have an Etherscan API key configured. Rotki etherscan queries will still work but will be very slow. If you are not using your own ethereum node, it is recommended to go to https://etherscan.io/registe
r, create an API key and then input it in the external service credentials setting of Rotki
[20/01/2021 18:48:17 CET] DEBUG rotkehlchen.externalapis.etherscan: Querying etherscan: https://api.etherscan.io/api?module=proxy&action=eth_call&to=0x3d9819210A31b4961b30EF54bE2aeD79B9c9Cd3B&data=0xbb82aa5e
KeyError: <rotkehlchen.user_messages.MessagesAggregator object at 0x7f44f2824c90>
                                                                                                                                        
The above exception was the direct cause of the following exception: 
                                                                                                                                        
Traceback (most recent call last):                                                                                                      
  File "src/gevent/greenlet.py", line 906, in gevent._gevent_cgreenlet.Greenlet.run             
  File "/home/lefteris/w/rotkehlchen/rotkehlchen/chain/ethereum/manager.py", line 304, in attempt_connect
    is_connected = web3.isConnected()  
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/web3/main.py", line 257, in isConnected
    return self.provider.isConnected()        
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/web3/providers/base.py", line 103, in isConnected
    response = self.make_request(RPCEndpoint('web3_clientVersion'), [])
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/web3/providers/rpc.py", line 95, in make_request
    **self.get_request_kwargs()                                                                                                         
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/web3/_utils/request.py", line 37, in make_post_request
    session = _get_session(endpoint_uri)                                                                                                
SystemError: PyEval_EvalFrameEx returned a result with an error set
2021-01-20T17:48:17Z <Greenlet at 0x7f44f7037950: <bound method EthereumManager.attempt_connect of <rotkehlchen.chain.ethereum.manager.EthereumManager object at 0x7f44f6f9c050>>(name=<NodeName.MYETHERWALLET: 6>, ethrpc_endpoint='https://nodes.mewapi.io/rpc/eth', mainnet_ch
eck=True)> failed with SystemError





../../.virtualenvs/rotkipy37/lib/python3.7/site-packages/requests/api.py:134: in put
    return request('put', url, data=data, **kwargs)
../../.virtualenvs/rotkipy37/lib/python3.7/site-packages/requests/api.py:61: in request
    return session.request(method=method, url=url, **kwargs)
../../.virtualenvs/rotkipy37/lib/python3.7/site-packages/requests/sessions.py:428: in __exit__
    self.close()
../../.virtualenvs/rotkipy37/lib/python3.7/site-packages/requests/sessions.py:747: in close
    v.close()
../../.virtualenvs/rotkipy37/lib/python3.7/site-packages/requests/adapters.py:325: in close
    self.poolmanager.clear()
../../.virtualenvs/rotkipy37/lib/python3.7/site-packages/urllib3/poolmanager.py:222: in clear
    self.pools.clear()
../../.virtualenvs/rotkipy37/lib/python3.7/site-packages/urllib3/_collections.py:100: in clear
    self.dispose_func(value)
../../.virtualenvs/rotkipy37/lib/python3.7/site-packages/urllib3/poolmanager.py:173: in <lambda>
    self.pools = RecentlyUsedContainer(num_pools, dispose_func=lambda p: p.close())
../../.virtualenvs/rotkipy37/lib/python3.7/site-packages/urllib3/connectionpool.py:490: in close
    conn = old_pool.get(block=False)
/usr/lib/python3.7/queue.py:181: in get
    self.not_full.notify()
/usr/lib/python3.7/threading.py:345: in notify
    if not self._is_owned():
/usr/lib/python3.7/threading.py:258: in _is_owned
    if self._lock.acquire(0):
../../.virtualenvs/rotkipy37/lib/python3.7/site-packages/gevent/thread.py:141: in acquire
    sleep()
../../.virtualenvs/rotkipy37/lib/python3.7/site-packages/gevent/hub.py:159: in sleep
    waiter.get()
src/gevent/_waiter.py:143: in gevent._gevent_c_waiter.Waiter.get
    ???
src/gevent/_waiter.py:154: in gevent._gevent_c_waiter.Waiter.get
    ???
src/gevent/_greenlet_primitives.py:61: in gevent._gevent_c_greenlet_primitives.SwitchOutGreenletWithLoop.switch
    ???
src/gevent/_greenlet_primitives.py:61: in gevent._gevent_c_greenlet_primitives.SwitchOutGreenletWithLoop.switch
    ???
src/gevent/_greenlet_primitives.py:65: in gevent._gevent_c_greenlet_primitives.SwitchOutGreenletWithLoop.switch
    ???
src/gevent/_gevent_c_greenlet_primitives.pxd:35: in gevent._gevent_c_greenlet_primitives._greenlet_switch
    ???
src/gevent/greenlet.py:906: in gevent._gevent_cgreenlet.Greenlet.run
    ???
rotkehlchen/chain/ethereum/manager.py:304: in attempt_connect
    is_connected = web3.isConnected()
../../.virtualenvs/rotkipy37/lib/python3.7/site-packages/web3/main.py:257: in isConnected
    return self.provider.isConnected()
../../.virtualenvs/rotkipy37/lib/python3.7/site-packages/web3/providers/base.py:103: in isConnected
    response = self.make_request(RPCEndpoint('web3_clientVersion'), [])
../../.virtualenvs/rotkipy37/lib/python3.7/site-packages/web3/providers/rpc.py:95: in make_request
    **self.get_request_kwargs()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

endpoint_uri = 'https://nodes.mewapi.io/rpc/eth', data = b'{"jsonrpc": "2.0", "method": "web3_clientVersion", "params": [], "id": 0}', args = ()
kwargs = {'headers': {'Content-Type': 'application/json', 'User-Agent': "Web3.py/5.15.0/<class 'web3.providers.rpc.HTTPProvider'>"}, 'timeout': 10}

    def make_post_request(endpoint_uri: URI, data: bytes, *args: Any, **kwargs: Any) -> bytes:
        kwargs.setdefault('timeout', 10)
>       session = _get_session(endpoint_uri)
E       SystemError: PyEval_EvalFrameEx returned a result with an error set

../../.virtualenvs/rotkipy37/lib/python3.7/site-packages/web3/_utils/request.py:37: SystemError
```

LefterisJP commented:
And another one

```
[20/01/2021 23:52:23 CET] DEBUG web3.providers.HTTPProvider: Making request HTTP. URI: https://web3.1inch.exchange, Method: web3_clientVersion
[20/01/2021 23:52:23 CET] DEBUG web3.providers.HTTPProvider: Making request HTTP. URI: https://nodes.mewapi.io/rpc/eth, Method: web3_clientVersion
[20/01/2021 23:52:23 CET] DEBUG web3.providers.HTTPProvider: Making request HTTP. URI: https://cloudflare-eth.com/, Method: web3_clientVersion
[20/01/2021 23:52:23 CET] WARNING rotkehlchen.user_messages: You do not have an Etherscan API key configured. Rotki etherscan queries will still work but will be very slow. If you are not using your own ethereum node, it is recommended to go to https://etherscan.io/registe
r, create an API key and then input it in the external service credentials setting of Rotki                                                                                                                                                                                      
[20/01/2021 23:52:23 CET] DEBUG rotkehlchen.externalapis.etherscan: Querying etherscan: https://api.etherscan.io/api?module=proxy&action=eth_call&to=0x3d9819210A31b4961b30EF54bE2aeD79B9c9Cd3B&data=0xbb82aa5e
KeyError: 'bc793c4bc2e7bf8d7a2ebcb2e6b6a57d'                                                                                            
                                                                    
The above exception was the direct cause of the following exception:       
                                                                                                                                        
Traceback (most recent call last):                                                                                                      
  File "src/gevent/greenlet.py", line 906, in gevent._gevent_cgreenlet.Greenlet.run                             
  File "/home/lefteris/w/rotkehlchen/rotkehlchen/chain/ethereum/manager.py", line 304, in attempt_connect
    is_connected = web3.isConnected()                                                                                                   
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/web3/main.py", line 257, in isConnected                  
    return self.provider.isConnected()                                                                                                  
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/web3/providers/base.py", line 103, in isConnected
    response = self.make_request(RPCEndpoint('web3_clientVersion'), [])                                                  
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/web3/providers/rpc.py", line 95, in make_request
    **self.get_request_kwargs()                                                                                                         
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/web3/_utils/request.py", line 37, in make_post_request
    session = _get_session(endpoint_uri)                                                                                                
SystemError: PyEval_EvalFrameEx returned a result with an error set                                                                     
2021-01-20T22:52:23Z <Greenlet at 0x7f3f080a2830: <bound method EthereumManager.attempt_connect of <rotkehlchen.chain.ethereum.manager.EthereumManager object at 0x7f3f080c3c90>>(name=<NodeName.MYETHERWALLET: 6>, ethrpc_endpoint='https://nodes.mewapi.io/rpc/eth', mainnet_ch
eck=True)> failed with SystemError                                                                                                      
                                                                                                                                        
F                                                                                                                                       
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> captured log >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
INFO     rotkehlchen.rotkehlchen:rotkehlchen.py:194 Unlocking user user=hania, create_new=True, sync_approval=yes, initial_settings=None 
DEBUG    rotkehlchen.accounting.accountant:accountant.py:58 Initializing Accountant                                 
DEBUG    rotkehlchen.exchanges.manager:manager.py:124 Initializing exchanges                                     
DEBUG    rotkehlchen.chain.ethereum.manager:manager.py:195 Initializing Ethereum Manager with own rpc endpoint: http://localhost:8545
DEBUG    rotkehlchen.chain.substrate.manager:manager.py:143 Initializing Kusama manager                         
WARNING  rotkehlchen.chain.substrate.manager:manager.py:159 Kusama manager won't attempt to connect to nodes connect_at_start=(<KusamaNodeName.OWN: 0>, <KusamaNodeName.PARITY: 1>), connect_on_startup=False, own_rpc_endpoint=http://localhost:9933
DEBUG    rotkehlchen.chain.manager:manager.py:213 Initializing ChainManager                                                     
DEBUG    rotkehlchen.rotkehlchen:rotkehlchen.py:309 User unlocking complete                                                        
DEBUG    rotkehlchen.api.rest:rest.py:128 Request successful response={"result": {"exchanges": [], "settings": {"have_premium": false, "version": 23, "last_write_ts": 0, "premium_should_sync": false, "include_crypto2crypto": true, "anonymized_logs": false, "last_data_uploa
d_ts": 0, "ui_floating_precision": 2, "taxfree_after_period": 31536000, "balance_save_frequency": 24, "include_gas_costs": true, "eth_rpc_endpoint": "http://localhost:8545", "ksm_rpc_endpoint": "http://localhost:9933", "main_currency": "USD", "date_display_format": "%d/%m/
%Y %H:%M:%S %Z", "last_balance_save": 0, "submit_usage_analytics": true, "kraken_account_type": "starter", "active_modules": ["makerdao_dsr", "makerdao_vaults", "aave", "compound", "yearn_vaults", "uniswap", "adex"], "frontend_settings": "", "account_for_assets_movements":
 true, "btc_derivation_gap_limit": 20, "calculate_past_cost_basis": true}}, "message": ""}, status_code=200             
DEBUG    web3.providers.HTTPProvider:rpc.py:90 Making request HTTP. URI: http://localhost:8545, Method: web3_clientVersion 
DEBUG    web3.providers.HTTPProvider:rpc.py:90 Making request HTTP. URI: https://api.mycryptoapi.com/eth, Method: web3_clientVersion
DEBUG    web3.providers.HTTPProvider:rpc.py:90 Making request HTTP. URI: https://mainnet-nethermind.blockscout.com/, Method: web3_clientVersion
DEBUG    web3.providers.HTTPProvider:rpc.py:90 Making request HTTP. URI: https://web3.1inch.exchange, Method: web3_clientVersion
DEBUG    web3.providers.HTTPProvider:rpc.py:90 Making request HTTP. URI: https://mainnet.eth.cloud.ava.do/, Method: web3_clientVersion
ERROR    rotkehlchen.db.dbhandler:dbhandler.py:249 At DB teardown could not open the DB: [Errno 2] No such file or directory: '/home/lefteris/.local/share/rotki/test_data/testuser/rotkehlchen.db'
ERROR    rotkehlchen.db.dbhandler:dbhandler.py:249 At DB teardown could not open the DB: [Errno 2] No such file or directory: '/home/lefteris/.local/share/rotki/test_data/testuser/rotkehlchen.db'

ERROR    rotkehlchen.db.dbhandler:dbhandler.py:249 At DB teardown could not open the DB: [Errno 2] No such file or directory: '/home/lefteris/.local/share/rotki/test_data/testuser/rotkehlchen.db'                                                                     [0/28798]
DEBUG    web3.providers.HTTPProvider:rpc.py:90 Making request HTTP. URI: https://web3.1inch.exchange, Method: web3_clientVersion                                                                                                                                                 
DEBUG    web3.providers.HTTPProvider:rpc.py:90 Making request HTTP. URI: https://nodes.mewapi.io/rpc/eth, Method: web3_clientVersion                                                                                                                                             
DEBUG    web3.providers.HTTPProvider:rpc.py:90 Making request HTTP. URI: https://cloudflare-eth.com/, Method: web3_clientVersion
WARNING  rotkehlchen.user_messages:user_messages.py:18 You do not have an Etherscan API key configured. Rotki etherscan queries will still work but will be very slow. If you are not using your own ethereum node, it is recommended to go to https://etherscan.io/register, cre
ate an API key and then input it in the external service credentials setting of Rotki                                                
DEBUG    rotkehlchen.externalapis.etherscan:etherscan.py:170 Querying etherscan: https://api.etherscan.io/api?module=proxy&action=eth_call&to=0x3d9819210A31b4961b30EF54bE2aeD79B9c9Cd3B&data=0xbb82aa5e
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> traceback >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
KeyError: 'bc793c4bc2e7bf8d7a2ebcb2e6b6a57d'                                                                                            
                                                                                                                                        
The above exception was the direct cause of the following exception:                                                                                                                                                                                                             
                                                                                                                                                                                                                                                                                 
rotkehlchen_api_server = <rotkehlchen.api.server.APIServer object at 0x7f3f0c4f7450>, data_dir = PosixPath('/home/lefteris/.local/share/rotki/test_data')                                                                                                                        
                                                                                                                                        
Fatal Python error: Segmentation fault                                                                                                  
                                                                                                                                        
Thread 0x00007f3f03e7f640 (most recent call first):                                                                                                                                                                                                                              
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/gevent/_threading.py", line 68 in wait                
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/gevent/_threading.py", line 158 in get                      
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/gevent/threadpool.py", line 187 in run                                                                                                                                                                 
                                                                                                                                                                                                                                                                                 
Thread 0x00007f3f08acf640 (most recent call first):                                                                                                                                                                                                                              
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/gevent/_threading.py", line 68 in wait                                                                                                                                                                 
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/gevent/_threading.py", line 158 in get                    
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/gevent/threadpool.py", line 187 in run                
                                                                                                                                                                                                                                                                                 
Current thread 0x00007f3f150a3740 (most recent call first):                                                                             
  File "<frozen importlib._bootstrap_external>", line 525 in _compile_bytecode                                                                                                                                                                                                   
  File "<frozen importlib._bootstrap_external>", line 857 in get_code                                                                                                                                                                                                            
  File "<frozen importlib._bootstrap_external>", line 724 in exec_module
  File "<frozen importlib._bootstrap>", line 677 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 967 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 983 in _find_and_load
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/pygments/lexers/__init__.py", line 18 in <module>                                                                                                                                                      
  File "<frozen importlib._bootstrap>", line 219 in _call_with_frames_removed
  File "<frozen importlib._bootstrap_external>", line 728 in exec_module
  File "<frozen importlib._bootstrap>", line 677 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 967 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 983 in _find_and_load                                                                      
  File "<frozen importlib._bootstrap>", line 219 in _call_with_frames_removed                                   
  File "<frozen importlib._bootstrap>", line 953 in _find_and_load_unlocked                                     
  File "<frozen importlib._bootstrap>", line 983 in _find_and_load
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/_pytest/_io/terminalwriter.py", line 202 in _highlight
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/_pytest/_io/terminalwriter.py", line 192 in _write_source
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/_pytest/_code/code.py", line 1101 in _write_entry_lines
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/_pytest/_code/code.py", line 1119 in toterminal
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/_pytest/_code/code.py", line 1025 in toterminal
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/_pytest/_code/code.py", line 995 in toterminal
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/_pytest/reports.py", line 87 in toterminal
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/_pytest/debugging.py", line 359 in _enter_pdb
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/_pytest/debugging.py", line 296 in pytest_exception_interact
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/pluggy/manager.py", line 87 in <lambda>
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/pluggy/hooks.py", line 286 in __call__           
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/_pytest/runner.py", line 221 in call_and_report
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/_pytest/runner.py", line 126 in runtestprotocol
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/_pytest/runner.py", line 109 in pytest_runtest_protocol
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/pluggy/manager.py", line 87 in <lambda>
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/_pytest/main.py", line 348 in pytest_runtestloop
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/pluggy/callers.py", line 187 in _multicall            
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/pluggy/manager.py", line 87 in <lambda>                  
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/pluggy/manager.py", line 93 in _hookexec               
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/pluggy/hooks.py", line 286 in __call__         
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/_pytest/main.py", line 323 in _main            
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/_pytest/main.py", line 269 in wrap_session    
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/_pytest/main.py", line 316 in pytest_cmdline_main
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/pluggy/callers.py", line 187 in _multicall   
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/pluggy/manager.py", line 87 in <lambda>                     
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/pluggy/manager.py", line 93 in _hookexec  
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/pluggy/hooks.py", line 286 in __call__ 
  File "/home/lefteris/.virtualenvs/rotkipy37/lib/python3.7/site-packages/_pytest/config/__init__.py", line 163 in main
  File "pytestgeventwrapper.py", line 8 in <module>                                                                                     
[1]    147402 segmentation fault (core dumped)  PYRE_PYTHON=1 python pytestgeventwrapper.py -xs rotkehlchen/tests/api --pdb  
```

LefterisJP commented:
Thinking more about it. I have no idea why the specific set of tests need to be given in the order they are, but they segfault reliably.

Now the why. That's another matter alltogether.

What I have noticed is that the test where it all goes down is the only test where a user is manually logged in properly via the API.

And that means that the bunch of connections to the various ethereum nodes from the ethereum manager are started there. And indeed from the log one can see the problem happens during those connections.

jamadden commented:
I was able to get this branch installed, and I think I understand the problem.

[web3 uses a package called lru-dict](https://github.com/ethereum/web3.py/blob/v5.15.0/setup.py#L83) to implement [its session cache](https://github.com/ethereum/web3.py/blob/v5.15.0/web3/_utils/request.py#L25) (where all the crashes occur).

[lru-dict](https://github.com/amitdev/lru-dict) is a C extension (that hasn't had a release in 4 years) that implements a least-recently-used scheme for caching, and it allows you to set a callback that's called when an old entry is evicted from the cache; web3 does this to close sessions as they are evicted:
```python
def _remove_session(key: str, session: requests.Session) -> None:
    session.close()

_session_cache = lru.LRU(8, callback=_remove_session)
```

However, `lru.LRU` is *not* reentrant. If the callback function mutates the dictionary in certain ways, the system crashes. For example, this script will crash 100% of the time:
```python
import lru

k = 0

def callback(*args):
    global k
    if k < 10:
        k += 1
        cache[k] = 'from callback'

cache = lru.LRU(5, callback)
for i in reversed(range(20)):
    cache[i] = i
```
```console
$ python /tmp/foo.py
Fatal Python error: Segmentation fault

Current thread 0x000000010cf69dc0 (most recent call first):
  File "/tmp/foo.py", line 9 in callback
  File "/tmp/foo.py", line 9 in callback
  File "/tmp/foo.py", line 9 in callback
  File "/tmp/foo.py", line 9 in callback
  File "/tmp/foo.py", line 9 in callback
  File "/tmp/foo.py", line 9 in callback
  File "/tmp/foo.py", line 9 in callback
  File "/tmp/foo.py", line 9 in callback
  File "/tmp/foo.py", line 9 in callback
  File "/tmp/foo.py", line 13 in <module>
Segmentation fault: 11
```

That's exactly what's happening here.

Something happens that ejects an entry from the cache, and we enter the web3 callback to close the session. However, calling `session.close()` can result in greenlets switching (one place this happens is that `session.close` tries to close the underlying `urllib3` objects, which acquires locks, and acquiring gevent locks can switch greenlets). If we then switch to a greenlet that attempts to add a new session to the cache, we re-enter the `lru.LRU` code in an unsafe state and we crash. 

Here's some example output showing this happening. I modified the callback to print a line when we enter ("Removing session"), and to print a line when we exit ("Removed session); you can see us entering it again from a different greenlet before we finish exiting it from the first greenlet. And then we crash.
```
>>> Storing session in <Greenlet at ...c30: <bound method EthereumManager.attempt_connect ...>
>>> Removing session in <Greenlet at ...c30: <bound method EthereumManager.attempt_connect ...>
  >>>> Allowing greenlet switch from <Greenlet at ...c30: <bound method EthereumManager.attempt_connect ...>
>>> Storing session in <Greenlet at ...9d0: <bound method EthereumManager.attempt_connect ...>
>>> Removing session in <Greenlet at ...9d0: <bound method EthereumManager.attempt_connect ...>
<<< Removed session in <Greenlet at ...9d0: <bound method EthereumManager.attempt_connect of ...>
 >>>> Allowing greenlet switch from  <Greenlet at ...c30: <bound method EthereumManager.attempt_connect ...>
<<<< Removed session in <Greenlet at ...c30: <bound method EthereumManager.attempt_connect of...>
KeyError: b'{"jsonrpc": "2.0", "method": "web3_clientVersion", "params": [], "id": 0}'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "src/gevent/greenlet.py", line 906, in gevent._gevent_cgreenlet.Greenlet.run
  File "/rotki/rotkehlchen/chain/ethereum/manager.py", line 305, in attempt_connect
    is_connected = web3.isConnected()
  File "//lib/python3.9/site-packages/web3/main.py", line 257, in isConnected
    return self.provider.isConnected()
  File "//lib/python3.9/site-packages/web3/providers/base.py", line 103, in isConnected
    response = self.make_request(RPCEndpoint('web3_clientVersion'), [])
  File "//lib/python3.9/site-packages/web3/providers/rpc.py", line 92, in make_request
    raw_response = make_post_request(
  File "//lib/python3.9/site-packages/web3/_utils/request.py", line 61, in make_post_request
    session = _get_session(endpoint_uri)
SystemError: _PyEval_EvalFrameDefault returned a result with an error set
```

In older versions of gevent, locks were less likely to allow greenlets to switch. But [beginning in 1.5a3](https://github.com/gevent/gevent/commit/06ede5d7d3dfb688022e3598e642761a36e22050) — literally the next version after the current version pin of 1.5a2 — that changed to make them more like native locks and allow greenlet switches more often.

So what to do about this? Well, you could argue that `web3._utils.request` is using its session cache in an unsafe manner. suspect you could probably get this exact same crash using regular threads instead of greenlets, because the `lru.LRU` object is *not* thread safe. They should have a lock around all uses of that object to be sure that there's only one callback in progress at a time. For example, this script typically crashes:
```python
import lru
import threading

def callback(key, value):
    with value: # Acquire a lock
        pass

cache = lru.LRU(5, callback)

def run():
    for i in range(100000):
        cache[i] = threading.Lock()

t1 = threading.Thread(target=run)
t2 = threading.Thread(target=run)

t1.start()
t2.start()
t1.join()
t2.join()
```
```console
$ python /tmp/foo_threads.py
Fatal Python error: Segmentation fault

Thread 0x000070000f21e000 (most recent call first):
  File "/tmp/foo.py", line 14 in run
  File "//3.9/lib/python3.9/threading.py", line 892 in run
  File "//3.9/lib/python3.9/threading.py", line 954 in _bootstrap_inner
  File "//3.9/lib/python3.9/threading.py", line 912 in _bootstrap

Current thread 0x000070000e21b000 (most recent call first):
  File "/tmp/foo.py", line 14 in run
  File "//3.9/lib/python3.9/threading.py", line 892 in run
  File "//3.9/lib/python3.9/threading.py", line 954 in _bootstrap_inner
  File "//3.9/lib/python3.9/threading.py", line 912 in _bootstrap

Thread 0x000000010f619dc0 (most recent call first):
  File "//3.9/lib/python3.9/threading.py", line 1049 in _wait_for_tstate_lock
  File "//3.9/lib/python3.9/threading.py", line 1033 in join
  File "/tmp/foo.py", line 21 in <module>
Segmentation fault: 11
```

Until then, you could monkey-patch that web3 module to use a safer data structure...

LefterisJP commented:
Hey @jamadden thank you. I really appreciate that you took the time to debug this issue with upgrading gevent and actually find the root cause of the problem and explain it so clearly.

I will take a look at your example scripts. Thank you.

> session.close() can result in greenlets switching (one place this happens is that session.close tries to close the underlying urllib3 objects, which acquires locks, and acquiring gevent locks can switch greenlets). If we then switch to a greenlet that attempts to add a new session to the cache, we re-enter the lru.LRU code in an unsafe state and we crash.

> In older versions of gevent, locks were less likely to allow greenlets to switch. But beginning in 1.5a3 — literally the next version after the current version pin of 1.5a2 — that changed to make them more like native locks and allow greenlet switches more often.

Very interesting! So this simply is a bug in web3.py that got uncovered due the new gevent versions also context switching at lock acquisitions. And the versions totally make sense. I tried even with 1.5a3 and I got the same error, but 1.5a2, the pinned one, works fine.

> Until then, you could monkey-patch that web3 module to use a safer data structure...

I think this would be the solution until https://github.com/ethereum/web3.py/issues/1847 is fixed and makes it into a release.

Again thank you! If there is any patreon, or any donation address to which I could "buy you a beer" so to say please don't hesitate to mention it. Can also be done in private: lefteris@rotki.com

LefterisJP commented:
> This pull request **introduces 1 alert** when merging [1ea58f0](https://github.com/rotki/rotki/commit/1ea58f056fdd18c490eca263ed14277de3296295) into [74ddf32](https://github.com/rotki/rotki/commit/74ddf32f15005520888fa62498739767b894b9d4) - [view on LGTM.com](https://lgtm.com/projects/g/rotki/rotki/rev/pr-339d7836ac83bdd45495d037fb79c7bc77a4e196)
> 
> **new alerts:**
> 
> * 1 for Unused import

Added an ignore for this in the code and it did not work. Won't bother to play with it and run tests again. Merging anyway since tests pass and also locally this solution I introduced here seems to have worked.
